proof_test.go

```diff
@@ -1124,3 +1125,690 @@ func Test_ProtoToProof(t *testing.T) {
+		{
+			currentStart:                0,
+			currentEnd:                  1,
+			subtreeRootMaximumLeafRange: 1,
+			expectedRange:               LeafRange{Start: 0, End: 1},
+		},
+		{
+			currentStart:                0,
+			currentEnd:                  16,
+			subtreeRootMaximumLeafRange: 16,
+			expectedRange:               LeafRange{Start: 0, End: 16},
+		},
+		{
+			currentStart:                0,
+			currentEnd:                  0,
+			subtreeRootMaximumLeafRange: 4,
+			expectError:                 true,
+		},
+		{
+			currentStart:                5,
+			currentEnd:                  2,
+			subtreeRootMaximumLeafRange: 4,
+			expectError:                 true,
+		},
+		{
+			currentStart:                5,
+			currentEnd:                  2,
+			subtreeRootMaximumLeafRange: 0,
+			expectError:                 true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(fmt.Sprintf("currentStart=%d, currentEnd=%d, subtreeRootMaximumLeafRange=%d", tt.currentStart, tt.currentEnd, tt.subtreeRootMaximumLeafRange), func(t *testing.T) {
+			result, err := nextLeafRange(tt.currentStart, tt.currentEnd, tt.subtreeRootMaximumLeafRange)
+			if tt.expectError {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+				assert.Equal(t, tt.expectedRange, result)
+			}
+		})
+	}
+}
+
+func TestSafeIntToUint(t *testing.T) {
+	tests := []struct {
+		input         int
+		expectedUint  uint
+		expectedError error
+	}{
+		{
+			input:         10,
+			expectedUint:  10,
+			expectedError: nil,
+		},
+		{
+			input:         0,
+			expectedUint:  0,
+			expectedError: nil,
+		},
+		{
+			input:         -5,
+			expectedUint:  0,
+			expectedError: fmt.Errorf("cannot convert a negative int %d to uint", -5),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(fmt.Sprintf("input=%d", tt.input), func(t *testing.T) {
+			result, err := safeIntToUint(tt.input)
+			if (err != nil) != (tt.expectedError != nil) || (err != nil && err.Error() != tt.expectedError.Error()) {
+				t.Errorf("expected error %v, got %v", tt.expectedError, err)
+			}
+			if result != tt.expectedUint {
+				t.Errorf("expected uint %v, got %v", tt.expectedUint, result)
+			}
+		})
+	}
+}
+
+func TestMinInt(t *testing.T) {
+	tests := []struct {
+		val1, val2 int
+		expected   int
+	}{
+		{
+			val1:     10,
+			val2:     20,
+			expected: 10,
+		},
+		{
+			val1:     -5,
+			val2:     6,
+			expected: -5,
+		},
+		{
+			val1:     5,
+			val2:     -6,
+			expected: -6,
+		},
+		{
+			val1:     -5,
+			val2:     -6,
+			expected: -6,
+		},
+		{
+			val1:     0,
+			val2:     0,
+			expected: 0,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(fmt.Sprintf("val1=%d, val2=%d", tt.val1, tt.val2), func(t *testing.T) {
+			result := minInt(tt.val1, tt.val2)
+			if result != tt.expected {
+				t.Errorf("expected %d, got %d", tt.expected, result)
+			}
+		})
+	}
+}
+
+func TestVerifySubtreeRootInclusion(t *testing.T) {
+	tree := exampleNMT(1, true, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15)
```

Comment:
Can we add in a test for a tree that isn't a power of two?
As this is all celestia specific, should we test with a subtreeWidth that follows adr13?

Reply:
> Can we add in a test for a tree that isn't a power of two?

The algorithm assumes a perfect tree... It won't work with a tree that isn't a power of 2

Reply:
> It won't work 

As in it panics or errors?